### PR TITLE
[ONNX] Add dynamic axes support to torchscript exporter with dynamo=True

### PR DESCRIPTION
**Pull request by titaiwangms**
This PR enables specific axe to be dynamic with calling torch.export.export and torch.export.Dim. 

Features:
(1) Turn dynamic_axes to dynamic_shapes
(2) Dim constraints remain the same (see test case with hitting constraints). This might give different user experience, since we didn't have any constraints in torchscript-onnx exporting.
(3) If input_names is used in dynamic_axes, ValueError will be raised, as input_names is currently not supported.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -512,33 +512,32 @@ def forward(self, x):
     """
 
     if dynamo:
+        if isinstance(model, (torch.jit.ScriptModule, torch.jit.ScriptFunction)):
+            raise ValueError(
```
Comment (justinchuby) [inline torch/onnx/utils.py:516]: TypeError

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -915,6 +914,60 @@ def _decide_input_format(model, args):
...
 
 
+@_beartype.beartype
+def _from_dynamic_axes_to_dynamic_shapes(
+    model,
+    dynamic_axes: Optional[
+        Union[Mapping[str, Mapping[int, str]], Mapping[str, Sequence[int]]]
+    ] = None,
+    input_names: Optional[Sequence[str]] = None,
+) -> Optional[Union[Dict[str, Any], Tuple[Any], List[Any]]]:
```
Comment (justinchuby) [inline torch/onnx/utils.py:924]: Is the output type correct?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -915,6 +914,60 @@ def _decide_input_format(model, args):
...
+    these will be converted to dynamic_shapes respectively:
+    (1) dynamic_shapes = {"x": {0: Dim("my_custom_axis_name_1")}, "y": {1: Dim("my_custom_axis_name_2")}}
+    (2) dynamic_shapes = {"x": {0: Dim("x_dim_0")}, "y": {1: Dim("y_dim_1")}}  # auto-generated dim names
+
+    """
+    if dynamic_axes is None:
+        return None
+    if input_names is None:
+        input_names = []
+    dynamic_shapes: Dict[str, Optional[Any]] = {}
```
Comment (justinchuby) [inline torch/onnx/utils.py:940]: Can the value type be refined?

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -915,6 +914,60 @@ def _decide_input_format(model, args):
...
+    for input_name, axes in dynamic_axes.items():
+        if input_name in input_names:
+            raise ValueError(
+                "input names is not supported yet. Please use model forward signature."
+            )
```
Comment (justinchuby) [inline torch/onnx/utils.py:945]: Perhaps explain the error a little bit more? Reading it briefly it isn't clear to me what is wrong and what needs to be done

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -915,6 +914,60 @@ def _decide_input_format(model, args):
...
+    (2) dynamic_shapes = {"x": {0: Dim("x_dim_0")}, "y": {1: Dim("y_dim_1")}}  # auto-generated dim names
+
+    """
+    if dynamic_axes is None:
+        return None
+    if input_names is None:
+        input_names = []
+    dynamic_shapes: Dict[str, Optional[Any]] = {}
+    for input_name, axes in dynamic_axes.items():
+        if input_name in input_names:
```
Comment (justinchuby) [inline torch/onnx/utils.py:942]: Would it be a good idea to turn input_names into a set for faster membership checks?

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -915,6 +914,60 @@ def _decide_input_format(model, args):
...
+        if isinstance(axes, dict):
+            dynamic_shapes[input_name] = {
+                k: torch.export.Dim(v) for k, v in axes.items()
+            }
+        elif isinstance(axes, list):
+            dynamic_shapes[input_name] = {
+                k: torch.export.Dim(f"{input_name}_dim_{k}") for k in axes
+            }
+        else:
+            raise ValueError(
```
Comment (justinchuby) [inline torch/onnx/utils.py:955]:
```suggestion
            raise TypeError(
```

File: test/onnx/dynamo/test_exporter_api.py [hunk 7]
```diff
@@ -271,17 +302,111 @@ def test_args_normalization_with_empty_dict_at_the_tail(self):
             onnx_program_from_old_exporter.model_proto,
         )
 
-    def test_dynamic_axes_enable_dynamic_shape(self):
+    def test_dynamic_axes_enable_dynamic_shapes_with_fully_specified_axes(self):
+        ep = torch.export.export(
```
Comment (justinchuby) [inline test/onnx/dynamo/test_exporter_api.py:306]: exported_program or program

File: test/onnx/dynamo/test_exporter_api.py [hunk 8]
```diff
@@ -271,17 +302,111 @@ def test_args_normalization_with_empty_dict_at_the_tail(self):
...
+                "b": {0: "customb_dim_0", 1: "customb_dim_1", 2: "customb_dim_2"},
+            },
+            dynamo=True,
+        )
+        self.assertEqual(
+            onnx_program_from_new_exporter.model_proto,
+            onnx_program_from_old_exporter.model_proto,
+        )
+
+    def test_dynamic_axes_enable_dynamic_shapes_with_default_axe_names(self):
```
Comment (justinchuby) [inline test/onnx/dynamo/test_exporter_api.py:344]: Just checking, this is testing the auto-None filling logic right?

File: test/onnx/dynamo/test_exporter_api.py [hunk 9]
```diff
@@ -271,17 +302,111 @@ def test_args_normalization_with_empty_dict_at_the_tail(self):
...
+                "b": [0, 1, 2],
+            },
+            dynamo=True,
+        )
+        self.assertEqual(
+            onnx_program_from_new_exporter.model_proto,
+            onnx_program_from_old_exporter.model_proto,
+        )
+
+    def test_dynamic_axes_enable_dynamic_shapes_with_static_axis(self):
```
Comment (justinchuby) [inline test/onnx/dynamo/test_exporter_api.py:383]: It would be helpful to include the expected behavior in the test name or find ways to make clear which part is static, as it is not entirely clear (to me just reading briefly)

**Review by justinchuby — COMMENTED** [review]
LGTM with minor comments. Thanks!

File: torch/onnx/utils.py (hunk 3 above)
Comment (titaiwangms) [reply torch/onnx/utils.py:944]: This is following https://github.com/pytorch/pytorch/blob/8733c4f4be47cc035fff339330492a36c2bfc702/torch/export/__init__.py#L78

File: torch/onnx/utils.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/utils.py:949]: For the time being, we only support:

```python

class Model(nn.module):
    def forward(self, input_x, input_y):
        ...

torch.onn.export(Model(), args, dynamic_axes={"input_x": ..., "input_y": ...}, dynamo=True)
```

input_names is not changing anything, so users have to use forward sig. Any suggestions?

File: test/onnx/dynamo/test_exporter_api.py (hunk 8 above)
Comment (titaiwangms) [reply test/onnx/dynamo/test_exporter_api.py:344]: Yes

File: test/onnx/dynamo/test_exporter_api.py (hunk 9 above)
Comment (titaiwangms) [reply test/onnx/dynamo/test_exporter_api.py:383]: How about test_dynamic_axes_enable_partial_dynamic_shapes?

Comment (justinchuby) [reply test/onnx/dynamo/test_exporter_api.py:383]: `enable` -> `supports`? lgtm

File: torch/onnx/utils.py [hunk 10]
```diff
@@ -512,33 +512,32 @@ def forward(self, x):
     """
 
     if dynamo:
+        if isinstance(model, (torch.jit.ScriptModule, torch.jit.ScriptFunction)):
+            raise TypeError(
+                "Dynamo export does not supported ScriptModule or ScriptFunction."
```
Comment (justinchuby) [inline torch/onnx/utils.py:517]: nit: grammar

**Review by justinchuby — APPROVED** [review]
I would proofread the error messages because they are user facing.

File: torch/onnx/utils.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/utils.py:949]: Got it. We can support that later. Thanks!
